masonry/src/widget/widget.rs

```diff
@@ -165,20 +167,42 @@ pub trait Widget: AsAny {
 
     // --- Auto-generated implementations ---
 
-    // FIXME
-    #[cfg(FALSE)]
-    /// Return which child, if any, has the given `pos` in its layout rect.
+    /// Return which child, if any, has the given `pos` in its layout rect and clip path. In case
+    /// of overlapping children, the last child as determined by [`Widget::children_ids`] is
+    /// chosen.
+    ///
+    /// The child returned is a direct child, not e.g. a grand-child.
     ///
-    /// The child return is a direct child, not eg a grand-child. The position is in
-    /// relative coordinates. (Eg `(0,0)` is the top-left corner of `self`).
+    /// Has a default implementation that can be overridden to search children more efficiently.
+    /// Custom implementations must uphold the conditions outlined above.
     ///
-    /// Has a default implementation, that can be overridden to search children more
-    /// efficiently.
-    fn get_child_at_pos(&self, pos: Point) -> Option<WidgetRef<'_, dyn Widget>> {
-        // layout_rect() is in parent coordinate space
-        self.children()
-            .into_iter()
-            .find(|child| child.state().layout_rect().contains(pos))
+    /// **pos** - the position in global coordinates (e.g. `(0,0)` is the top-left corner of the
+    /// window).
+    fn get_child_at_pos<'c>(
+        &self,
+        ctx: &'c QueryCtx,
+        pos: Point,
+    ) -> Option<WidgetRef<'c, dyn Widget>> {
+        // Assumes `Self::children_ids` is in increasing "z-order", picking the last child in case
+        // of overlapping children.
+        for child_id in self.children_ids().iter().rev() {
+            let child = ctx.get(*child_id);
+
+            let relative_pos = pos - child.state().window_origin().to_vec2();
+            // The position must be inside the child's layout and inside the child's clip path (if
+            // any).
+            if !child.widget.skip_pointer()
+                && child.state().window_layout_rect().contains(pos)
+                && child
+                    .state()
+                    .clip
+                    .map_or(true, |clip| clip.contains(relative_pos))
```

Comment:
The clip check should be outside the loop. The logic (for now) is that a widget's clip only constrains its children.

(We might change it, because it *is* counter-intuitive.)

Reply:
Ah, right. I indeed thought it affected the widget itself. That means the clip check can be removed from `WidgetRef::find_widget_at_pos`, as it's now done by `Widget::get_child_at_pos`.